_posts/2025-04-22-test-classloading-rewrite.adoc

```diff
@@ -0,0 +1,63 @@
+---
+layout: post
+title: 'The internals (and a few externals) of Quarkus test classloading have changed'
+tags: announcement testing
+synopsis: 'The way that Quarkus loads test classes has been updated. Most tests will not need to change, but here are some things to watch out for.'
+author: hcummins
+---
+
+== What's changing?
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
+Other test extensions haven't changed, nor has dev mode and production mode classloading.
+This change should make Quarkus testing work better, and it allowed us to fix a pile of longstanding bugs.
+However, it did introduce a few bugs we know about, and most likely also some bugs we don't yet know about.
+We're keen to get feedback from the community so that we can get fixing.
+
+== Why?
+
+In previous versions, Quarkus tests were invoked using the default JUnit classloader, and then executed in a different, Quarkus-aware, classloader.
+
+This mostly worked very well, and meant that `QuarkusTest` tests mostly behaved as if they were part of the same application as the code under test.
+The Quarkus test framework could start and stop Quarkus instances at the right point in the test lifecycle, inject CDI dependencies, and do other useful Quarkus bytecode manipulation.
+However, some use cases didn't work. Tests using advanced JUnit features like `@TestTemplate` and `@ParameterizedTest` sometimes found that the same test code might appear to run in several classloaders in a single test, or that injected dependencies weren't always available.
+
+While Quarkus extensions can do all sorts of marvellous bytecode manipulation to improve the developer experience, they cannot manipulate test classes with the same freedom that they do normal application classes.
+
+Over time, test-related defects were building up that couldn't be changed without a fundamental rewrite of test classloading. The Quarkus test code itself was also growing ever-more complex as it tried to work around various JUnit edge cases. Moving test instances from one classloader to another involved serializing and deserialization, which is harder to implement on newer JVM versions with tighter class security. For example, Quarkus used to use XStream as the serialization provider, but XStream no longer works with Java 17 and higher, because of reflection restrictions in the newer JVMs.
+
+What if, instead, Quarkus tests were simply run in the same classloader used to to load them?
+
+== What you need to do
+
+From Quarkus 3.22 onwards, this is exactly how `@QuarkusTest` classloading works.
+What do your tests need to change in order to work with the new architecture?
+Hopefully, nothing.
```

Comment:
I would write it: 

**nothing** (hopefully)

But it reduces the level of confidence.